SOQL code-completion on WHERE clause expressions

dobladez opened the pull request:
### What does this PR do?

Code completion of WHERE clause expressions
    
  * Complete operators based on type of field
  * Complete literal values based on type of field
  * Show type name (and icon) on completion items for fields
  * Offer NULL as a completion value only if field is nillable
    
Also: refactoring of "placeholder" completion items: receive JSON data instead
of parseable values inside item label string.

### What issues does this PR fix or reference?
@W-8597428@

Related LSP-server PR: https://github.com/forcedotcom/soql-tooling/pull/112

### Functionality Before
n/a

### Functionality After
https://user-images.githubusercontent.com/152839/103246284-b139f700-4941-11eb-9228-2a56fc06dd87.mov

https://user-images.githubusercontent.com/152839/103816838-60fb1d00-5044-11eb-9c42-8336da6bb86b.mov

lcampos commented on packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts:
```diff
@@ -0,0 +1,317 @@
+
+      const handler = expandFunctions[commandName];
+      if (!handler) {
+        console.log(`Unknown SOQL completion command ${commandName}!`);
```
Not sure if this needs to be surfaced to the customer but we should at least track it with the telemetry service. Same for the rest of the `console.log`s in the rest of this file.

dobladez replied:
Good point. I'll add telemetry

dobladez replied:
Done

lcampos commented on packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts:
```diff
@@ -0,0 +1,317 @@
+    try {
+      const objMetadata = await retrieveSObject(soqlContext.sobjectName);
+      const sobjectFields = objMetadata.fields.reduce((fieldItems, field) => {
+        const icon = iconsByFieldType[field.type];
```
Not sure I understand the purpose of these icons. Is it just for decoration or is there a functional purpose for using them ?

dobladez replied:
It shows the user the type of each field.  I personally don't like the icons, but do find the information useful. It could be text instead.

In addition to the icons (which are currently displayed on all items of the list), I've included the textual name of the type to the right of the selected item only. You can see it on the videos above, but here's a screenshot example:
![image](https://user-images.githubusercontent.com/152839/104022187-b8150500-519e-11eb-94f6-f053705378ed.png)

brpowell replied:
I've never seen a language server use emojis and it could be confusing to decode for users (as much as i love using emojis 🎉 ). I think we should just stick with displaying the type via the CompletionItem API like you have already.

lcampos replied:
@dobladez let's get rid of them, as @brpowell pointed out this is not how the rest of the language servers display information and we want to be consistent with the rest since this will eventually be used in Apex code.

dobladez replied:
done

lcampos commented on packages/salesforcedx-vscode-soql/src/client/queryValidation.ts:
```diff
@@ -0,0 +1,42 @@
+        return { result: queryData };
+      });
+    } catch (e) {
+      // NOTE: The return value must be serializable, for JSON-RPC.
```
Should we be logging (telemetry) this too or is there a place where this eventually gets logged already ?

dobladez replied:
That particular error is displayed to the user.
The comment is there to avoid the temptation of replacing that return call with `return {error: e}`, which seems natural but fails due to `e` not guaranteed to be JSON-serializable.

praksb commented on packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts:
```diff
@@ -0,0 +1,317 @@
+    soqlContext?: SoqlItemContext
+  ) => Promise<ProtocolCompletionItem[]>;
+} = {
+  SOBJECTS_PLACEHOLDER: async (
```
expandFunctions is now doing lot more things that I originally thought it would do (which is just expand sobjects and its fields). 

I feel all this logic should move over to the language server instead of the client. I think we just go ahead and implement what we discussed in the last meeting to have the language server take in something like an sObjectProvider and move all the code here to the language server and keep the client simple.

praksb replied:
Also, SOQL completion in Apex would rely on the SOQL language server and we cannot have this logic duplicated over there as well in the Apex extension.

dobladez replied:
Agreed.  We have to define how to do it though.... it's not trivial and we need to agree on an approach.

BTW: as an intermediate solution, some of this logic can be moved to the server-side. For example, for this new logic for type-aware fields, the server could send all completion items with extra metadata indicating for which types the item applies... so the client would only (and optionally) filter them based on the field type. Still not a solution, but wanted to mentioning it.

praksb reviewed (changes requested):
Let's talk during our dev hangout. I feel that the placeholder expansion logic is not simple anymore and it needs to be switch over to the language server.

brpowell commented on packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts:
```diff
@@ -0,0 +1,320 @@
+
+      return sobjectItems;
+    } catch (metadataErrors) {
+      return [];
```
If `retrieveSObjects()` throws an error, seems like something we might want to report

dobladez replied:
Thanks. Done.

brpowell commented on packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts:
```diff
@@ -0,0 +1,320 @@
+      }, [] as ProtocolCompletionItem[]);
+      return sobjectFields;
+    } catch (metadataError) {
+      return [];
```
Similar thing here, how do we get more insight into what the thrown issue is?

dobladez commented:
Ready to re-review:
* Removed type icons
* Added telemetry
* Added better error handling when fetching objects metadata

In the future, we will move all logic to the LSP server-side. First we need to define an approach in coordination with the other modules and extensions that fetch and cache org data.

AnanyaJha commented on packages/salesforcedx-vscode-soql/test/vscode-integration/client/completion.test.ts:
```diff
@@ -5,18 +5,23 @@
  * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
 import { expect } from 'chai';
+import { Connection } from 'jsforce';
```
Do we need to use the 'Connection' from jsforce specifically? We typically access the 'Connection' object from `@salesforce/core`

dobladez replied:
Changed. Thanks!
